src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/LockFactory.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal class LockFactory
```

Comment:
I've been going around in my head a bit about the name of this class. It's not really a factory for creating locks; it's just a convenience wrapper around a single lock with various methods to enter that lock.

Reply:
With the methods named `GetXXX` I think it made more sense, but changing to `EnterXXX` what do you think about just calling `ReadWriteLockWrapper`?

Reply:
Sure, or even `ReadWriterLocker`.